v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -75,32 +76,32 @@ func (c *Cleaner) Run(ctx context.Context) error {
 			return errors.New(fmt.Sprintf("it doesn't look like your version of ASO is one that supports deprecating version %q. Have you upgraded ASO yet?", deprecatedVersion))
 		}
 
-		// Make sure to use a version that hasn't been deprecated for migration. Deprecated versions will not be in our
-		// scheme, and so we cannot List/PUT with them. Instead, use the next available storedVersion.
-		activeVersion := newStoredVersions[0]
-
-		// If we found an updated slice, which implies, we have found a version to deprecate. Then only continue with the cleaning process
-		if len(newStoredVersions) != len(crd.Status.StoredVersions) {
-			klog.Infof("Starting cleanup for %q", crd.Name)
-			objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, activeVersion)
-			if err != nil {
-				return err
-			}
-
-			err = c.migrateObjects(ctx, objectsToMigrate)
-			if err != nil {
-				return err
-			}
-
-			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
-			if err != nil {
-				return err
-			}
-
-			updated++
-		} else {
+		// If the slice was not updated, there is no version to deprecate.
+		if len(newStoredVersions) == len(crd.Status.StoredVersions) {
 			klog.Infof("Nothing to update for %q\n", crd.Name)
+			continue
+		}
+
+		// Make sure to use a version that hasn't been deprecated for migration. Deprecated versions will not be in our
+		// scheme, and so we cannot List/PUT with them. Instead, use the next available version.
+		activeVersion := getVersionFromStoredVersion(newStoredVersions[0])
```

Comment:
Not sure that this is the right one to use. 

If `newStoredVersions` is in non-deterministic order, this runs the risk of using a version that is missing some properties, and thus experiencing data-loss.

If `newStoredVersions` is in oldest-to-newest order, then selecting the oldest one is almost guaranteed to cause data loss because properties are unknown. 

Or maybe the result will be that `asoctl` will fail?

Either way, not good. Need to proactively select the latest non-preview version.

Reply:
To clarify the problem.

Say we currently have support for three different ARM API versions:
* `v2020-01-01`
* `v2021-01-01` - introduces property `Foo`
* `v2022-01-01` - introduces property `Bar`

We'll therefore have the following ASO CRD versions:
* `v1alpha1api20200101`
* `v1alpha1api20210101`
* `v1alpha1api20220101`
* `v1beta20200101`
* `v1beta20210101`
* `v1beta20220101`

Now suppose we have a `v1alpha1api20220101` resource which is using both the `Foo` and `Bar` properties.

When doing the upgrade, `newStoredVersions[0]` will be `v1beta20200101` - this version doesn't have `Foo` nor `Bar`, so it's not possible to save the amended resource as that version; I think the save will then fail because the api-server doesn't know what to do with the extra properties.